servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherZipper.java

```diff
@@ -0,0 +1,302 @@
+                subscriber.onSubscribe(subscription);
+            }
+
+            @Override
+            public void onNext(@Nullable final ZipArg zipArg) {
+                assert zipArg != null;
+                array[zipArg.index].add(wrapNull(zipArg.value));
+                nonEmptyQueueIndexes |= 1L << zipArg.index;
+                if (nonEmptyQueueIndexes == ALL_NON_EMPTY_MASK) {
+                    Object[] zipArray = new Object[array.length];
+                    for (int i = 0; i < array.length; ++i) {
+                        final Queue<Object> arrayQueue = array[i];
+                        final Object queuePoll = arrayQueue.poll();
+                        assert queuePoll != null;
+                        if (arrayQueue.isEmpty()) {
+                            nonEmptyQueueIndexes &= ~(1L << i);
+                        }
+                        zipArray[i] = unwrapNullUnchecked(queuePoll);
+                        // Allow this subscriber to request more if demand is pending.
+                        // Reentry: note that we call out to request more before we dequeued the current set of signals
+                        // which in theory may result in out of order delivery. However, flatMap protects against
+                        // reentry so no need to provide double protection in this method.
+                        demandSubscribers[i].incrementSourceEmitted();
+                    }
+                    subscriber.onNext(zipper.apply(zipArray));
+                } else {
+                    assert subscription != null;
+                    subscription.request(1);
+                }
+            }
+
+            @Override
+            public void onError(final Throwable t) {
+                subscriber.onError(t);
+            }
+
+            @Override
+            public void onComplete() {
+                List<Integer> nonEmptyIndexes = new ArrayList<>();
+                for (int i = 0; i < array.length; ++i) {
+                    if ((nonEmptyQueueIndexes & (1L << i)) != 0) {
+                        nonEmptyIndexes.add(i);
+                    }
+                }
+                if (nonEmptyIndexes.isEmpty()) {
+                    subscriber.onComplete();
+                } else {
+                    StringBuilder sb = new StringBuilder(20 + 68 + nonEmptyIndexes.size() * 4);
+                    sb.append("Publisher indexes: [");
+                    Iterator<Integer> itr = nonEmptyIndexes.iterator();
+                    sb.append(itr.next()); // safe to call next(), already checked is not empty.
+                    while (itr.hasNext()) {
+                        sb.append(", ").append(itr.next());
+                    }
+                    sb.append("] had onNext signals queued when onComplete terminal signal received");
+                    subscriber.onError(new IllegalStateException(sb.toString()));
+                }
+            }
```

Comment:
I don't think this is the same behavior described in https://reactivex.io/documentation/operators/zip.html. It suggests that its okay to have extra signals emitted and they are simply dropped which is the behavior of most collection zip operators that I've seen. At a minimum, the behavior should be documented so users know they should have publishers of all the same length, which I think could be non-trivial in some cases.

Along these same lines, I don't think this implementation will ever terminate if one of the suppliers is an infinite stream. It seems pretty reasonable to have one source that is an infinite stream an the join it with another finite stream and expect a finite result.